SofaKernel/modules/SofaExplicitOdeSolver/src/SofaExplicitOdeSolver/EulerSolver.cpp

```diff
@@ -53,127 +52,190 @@ EulerExplicitSolver::EulerExplicitSolver()
+        // In our case, the operations will be executed to compute the new velocity vector,
+        // and the new position vector. The order of execution is defined by
+        // the symplectic property of the solver.
+        VMultiOp ops(2);
+
+        // Change order of operations depending on the symplectic flag
+        const VMultiOp::size_type posId = d_symplectic.getValue(); // 1 if symplectic, 0 otherwise
+        const VMultiOp::size_type velId = 1 - posId; // 0 if symplectic, 1 otherwise
+
+        // Access the set of operations corresponding to the velocity vector
+        // In case of symplectic solver, these operations are executed first.
+        auto& ops_vel = ops[velId];
+
+        // Associate the new velocity vector as the result to this set of operations
+        ops_vel.first = newVel;
+
+        // The two following operations are actually a unique operation: newVel = vel + dt * acc
+        // The value 1.0 indicates that the first operation is based on the values
+        // in the second pair and, therefore, the second operation is discarded.
+        ops_vel.second.emplace_back(vel.id(), 1.0);
+        ops_vel.second.emplace_back(acc.id(), dt);
+
+        // Access the set of operations corresponding to the position vector
+        // In case of symplectic solver, these operations are executed second.
+        auto& ops_pos = ops[posId];
+
+        // Associate the new position vector as the result to this set of operations
+        ops_pos.first = newPos;
+
+        // The two following operations are actually a unique operation: newPos = pos + dt * v
+        // where v is "newVel" in case of a symplectic solver, and "vel" otherwise.
+        // If symplectic: newPos = pos + dt * newVel, executed after newVel has been computed
+        // If not symplectic: newPos = pos + dt * vel
+        // The value 1.0 indicates that the first operation is based on the values
+        // in the second pair and, therefore, the second operation is discarded.
+        ops_pos.second.emplace_back(pos.id(), 1.0);
+        ops_pos.second.emplace_back(d_symplectic.getValue() ? newVel.id() : vel.id(), dt);
+
+        // Execute the defined operations to compute the new velocity vector and
+        // the new position vector.
+        // 1. Calls the "vMultiOp" method of every mapped BaseMechanicalState objects found in the
+        // current context tree. This method may be called with different parameters than for the non-mapped
+        // BaseMechanicalState objects.
+        // 2. Calls the "vMultiOp" method of every BaseMechanicalState objects found in the
+        // current context tree.
+        vop->v_multiop(ops);
+
+        // Calls "solveConstraint" on every ConstraintSolver objects found in the current context tree.
+        mop->solveConstraint(newVel,core::ConstraintParams::VEL);
+        mop->solveConstraint(newPos,core::ConstraintParams::POS);
```

Comment:
I noticed that too, but I don't know the answer. @ChristianDuriez can you help us on this question?

Reply:
Short answer: it depends on how the constraint is implemented.

In the non-optimized version, constraints are applied on the velocity, and then the position is updated. Therefore, the new position takes into account the changes due to the constraints on the velocity. Finally, constraints on the position are applied.
In the optimized version, the position is updated first, and then the constraints on the velocity are applied. Therefore, the new position does not take into account the constraints on the velocity. However, it may have the same effect, if  the constraint applied on the position has the same effect than the constraints applied on the velocity.
To be completely rigorous, I think the non-optimized version is the good one. However, in practice the optimized version works similarly to the non-optimized version, because constraints on velocity and position are consistent.
Let's see if it worths keeping the "optimized" version after a benchmark.